shell/platform/fuchsia/flutter/software_surface.h

```diff
@@ -0,0 +1,115 @@
+// Copyright 2013 The Flutter Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+#pragma once
+
+#include <fuchsia/sysmem/cpp/fidl.h>
+#include <fuchsia/ui/composition/cpp/fidl.h>
+#include <lib/async/cpp/wait.h>
+#include <lib/ui/scenic/cpp/resources.h>
+#include <lib/zx/event.h>
+#include <lib/zx/vmo.h>
+
+#include <array>
+#include <cstdint>
+#include <memory>
+
+#include "flutter/fml/macros.h"
+#include "third_party/skia/include/core/SkSize.h"
+#include "third_party/skia/include/core/SkSurface.h"
+
+#include "surface_producer.h"
+
+namespace flutter_runner {
+
+class SoftwareSurface final : public SurfaceProducerSurface {
+ public:
+  SoftwareSurface(fuchsia::sysmem::AllocatorSyncPtr& sysmem_allocator,
+                  fuchsia::ui::composition::AllocatorPtr& flatland_allocator,
+                  scenic::Session* session,
+                  const SkISize& size,
+                  uint32_t buffer_id);
+
+  ~SoftwareSurface() override;
+
+  size_t GetAllocationSize() const { return 0; }
+
+  // |SurfaceProducerSurface|
+  size_t AdvanceAndGetAge() override;
+
+  // |SurfaceProducerSurface|
+  bool FlushSessionAcquireAndReleaseEvents() override;
+
+  // |SurfaceProducerSurface|
+  bool IsValid() const override;
+
+  // |SurfaceProducerSurface|
+  SkISize GetSize() const override;
+
+  // Note: It is safe for the caller to collect the surface in the
+  // |on_writes_committed| callback.
+  void SignalWritesFinished(
+      const std::function<void(void)>& on_writes_committed) override;
+
+  // |SurfaceProducerSurface|
+  void SetImageId(uint32_t image_id) override;
+
+  // |SurfaceProducerSurface|
+  uint32_t GetImageId() override;
+
+  // |SurfaceProducerSurface|
+  sk_sp<SkSurface> GetSkiaSurface() const override;
+
+  // |SurfaceProducerSurface|
+  fuchsia::ui::composition::BufferCollectionImportToken
+  GetBufferCollectionImportToken() override;
+
+  // |SurfaceProducerSurface|
+  zx::event GetAcquireFence() override;
+
+  // |SurfaceProducerSurface|
+  zx::event GetReleaseFence() override;
+
+  // |SurfaceProducerSurface|
+  void SetReleaseImageCallback(
+      ReleaseImageCallback release_image_callback) override;
+
+ private:
+  void OnHandleReady(async_dispatcher_t* dispatcher,
+                     async::WaitBase* wait,
+                     zx_status_t status,
+                     const zx_packet_signal_t* signal);
+
+  bool SetupSkiaSurface(
+      fuchsia::sysmem::AllocatorSyncPtr& sysmem_allocator,
+      fuchsia::ui::composition::AllocatorPtr& flatland_allocator,
+      const SkISize& size,
+      uint32_t buffer_id);
+
+  bool CreateFences();
+
+  void PushSessionImageSetupOps(scenic::Session* session);
+
+  void Reset();
+
+  scenic::Session* session_;
+  uint32_t image_id_ = 0;
```

Comment:
Should we add a docstring explaining what "space" this ID is in, i.e. where it needs to come from/be used?

Reply:
Even better, this can be made a scenic::ResourceId which makes the type/space obvious.